In [1]:
from konlpy.tag import Mecab
import pandas as pd
import numpy as np
import re
import time

from sklearn.feature_extraction.text import CountVectorizer

In [2]:
class review_anal:        
    
    # 리뷰(csv파일) 리스트로 가져오기
    def get_review(review):
    
        raw_review = pd.read_csv(review, encoding='utf-8')['Content']
    
        # 리뷰 텍스트 리스트에 담기
        raw_review_list=[]
        hangeul = re.compile('[^ ㄱ-ㅣ가-힣]+')
        for i in range(len(raw_review)):
            try:
                raw_review_list.append(hangeul.sub('',raw_review.loc[i]))
            except:
                pass
        return raw_review_list    
    
    
    # 리뷰 토큰화 전
    def token_review(
#                         ngram_range="(1,1)"
#                         , tokenizer=
                    ):
        # stop_words
        stop_words = set(['은','는','이','가','하','아','것','들','의','있','그','되','수','보','주','등','한'
                          ,'나','같','우리','때','년','지','오','말','일','그렇','위하','때문','그것'
                          ,'두','말하','알','그러나','받','일','또','그런','문제','그리고','크','중','나오','가지'
                          ,'시키','지금','생각하','그러','속','하나','집','살','적','월','데','자신','안','어떤'
                          ,'내','경우','명','생각','시간','그녀','다시','이런','앞','보이','번','나','다른','어떻','전','들'
                          ,'사실','이렇','점','싶','말','정도','좀','원','통하','소리','놓'
                          ,'error','으로','잇님','잇님들'])
    
        # BOW : 토큰-벡터
        mecab = Mecab()
        vect = CountVectorizer(
                                ngram_range=(2,3)
                                , tokenizer=mecab.morphs
                                , stop_words=stop_words
#                                 , max_df=   # 0.0~1.0 사이 실수(%) 혹은 정수(개수)
                                , min_df = 25  # 0.0~1.0 사이 실수(%) 혹은 정수(개수)
#                                 , max_feature=   # 추출할 피쳐의 개수 지정(정수)
                              )    
        return vect
    
    
    
    
    # fit, 어휘사전
    def vocabulary_review(review):
        vect = review_anal.token_review()
        vect.fit(review_anal.get_review(review))
        voca = vect.vocabulary_
        return voca
        
        
    # bag_of_words_list    
    def bow_review(review):
        vect = review_anal.token_review()
#         fitting = review_anal.vocabulary_review(review)
        vect.fit(review_anal.get_review(review))
        
        bag_of_words = vect.transform(review_anal.get_review(review))
        bag_of_words_list = bag_of_words.toarray()
        
        return bag_of_words_list
    
    
    
    # bag_of_words_names_list
    def names_review(review):
        vect = review_anal.token_review()
        vect.fit(review_anal.get_review(review))
        
        bag_of_words_names = vect.get_feature_names()
        
        return bag_of_words_names

In [3]:
%%time

names_review_result= review_anal.names_review('./post_crawling_통합_정제_2.csv')

df = pd.DataFrame({
    'names' : names_review_result
})

df.to_csv('./post_crawling_통합_정제_2_names.csv', encoding='utf-8')

/home/komei54/anaconda3/envs/text-env/lib/python3.6/site-packages/sklearn/feature_extraction/text.py:300: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['님', '잇'] not in stop_words.
  'stop_words.' % sorted(inconsistent))


CPU times: user 1h 6min 13s, sys: 1min 17s, total: 1h 7min 31s
Wall time: 1h 7min 33s


In [4]:
df13_names = pd.read_csv('./post_crawling_통합_정제_2_names.csv')
df13_names

Unnamed: 0   names
0                 0     ㄱ ㄱ
1                 1   ㄱ ㄱ ㄱ
2                 2   ㄱ ㄱ ㅅ
3                 3   ㄱ ㄱ ㅋ
4                 4  ㄱ ㄱ ㅋㅋ
...             ...     ...
1357173     1357173     힣 ㅎ
1357174     1357174    힣 ㅎㅎ
1357175     1357175   힣 ㅎㅎㅎ
1357176     1357176     힣 히
1357177     1357177     힣 힣

[1357178 rows x 2 columns]

In [5]:
for a in df13_names.names[40000:45000]:
    print(a, end=", ")

감사 함, 감사 함 과, 감사 함 니, 감사 함 돠, 감사 함 을, 감사 함다, 감사 함당, 감사 함미, 감사 함미 다, 감사 합, 감사 합 니, 감사 합 니당, 감사 합니다, 감사 합니다 ㅋ, 감사 합니다 ㅋㅋ, 감사 합니다 ㅋㅋㅋ, 감사 합니다 ㅎ, 감사 합니다 ㅎㅎ, 감사 합니다 ㅎㅎㅎ, 감사 합니다 ㅠㅠ, 감사 합니다 감사, 감사 합니다 건강, 감사 합니다 고, 감사 합니다 고객, 감사 합니다 고맙, 감사 합니다 공감, 감사 합니다 그냥, 감사 합니다 그래서, 감사 합니다 그럼, 감사 합니다 근데, 감사 합니다 글, 감사 합니다 꾸벅, 감사 합니다 끝, 감사 합니다 너무, 감사 합니다 늘, 감사 합니다 다, 감사 합니다 다음, 감사 합니다 대전, 감사 합니다 더, 감사 합니다 덕분, 감사 합니다 도, 감사 합니다 도움, 감사 합니다 라고, 감사 합니다 를, 감사 합니다 마지막, 감사 합니다 맛있, 감사 합니다 먹, 감사 합니다 모두, 감사 합니다 모든, 감사 합니다 문, 감사 합니다 복, 감사 합니다 본, 감사 합니다 부산, 감사 합니다 블로그, 감사 합니다 사랑, 감사 합니다 사장, 감사 합니다 사진, 감사 합니다 서울, 감사 합니다 수원, 감사 합니다 아래, 감사 합니다 아이, 감사 합니다 어, 감사 합니다 언제나, 감사 합니다 에, 감사 합니다 여기, 감사 합니다 여러분, 감사 합니다 오늘, 감사 합니다 요즘, 감사 합니다 이렇게, 감사 합니다 이번, 감사 합니다 이상, 감사 합니다 이웃, 감사 합니다 이제, 감사 합니다 인, 감사 합니다 인사, 감사 합니다 인천, 감사 합니다 잘, 감사 합니다 재생, 감사 합니다 저, 감사 합니다 저희, 감사 합니다 정말, 감사 합니다 제, 감사 합니다 제주, 감사 합니다 좋, 감사 합니다 주문, 감사 합니다 주소, 감사 합니다 즐거운, 감사 합니다 진짜, 감사 합니다 참, 감사 합니다 픽스, 감사 합니다 하루, 감사 합니다 항상, 감사 합니다 행복, 감사 합니다 화면, 감사 합니다요, 감사 해, 감사 해 며, 

갔 다 동네, 갔 다 동생, 갔 다 동안, 갔 다 둘, 갔 다 드디어, 갔 다 들어가, 갔 다 딱, 갔 다 라고, 갔 다 를, 갔 다 마, 갔 다 마지막, 갔 다 마침, 갔 다 만, 갔 다 맛, 갔 다 맛있, 갔 다 맨날, 갔 다 먹, 갔 다 먼저, 갔 다 메뉴, 갔 다 며, 갔 다 면, 갔 다 면서, 갔 다 몇, 갔 다 무슨, 갔 다 물론, 갔 다 뭐, 갔 다 뭔가, 갔 다 바로, 갔 다 밤, 갔 다 밥, 갔 다 배, 갔 다 버스, 갔 다 보다, 갔 다 부산, 갔 다 부천, 갔 다 분, 갔 다 분위기, 갔 다 블로그, 갔 다 비, 갔 다 사, 갔 다 사람, 갔 다 사랑, 갔 다 사진, 갔 다 서, 갔 다 서울, 갔 다 세, 갔 다 세요, 갔 다 셔서, 갔 다 셨, 갔 다 솔직히, 갔 다 수원, 갔 다 숙소, 갔 다 술, 갔 다 시, 갔 다 시장, 갔 다 식당, 갔 다 신, 갔 다 아니, 갔 다 아빠, 갔 다 아이, 갔 다 아주, 갔 다 아직, 갔 다 아침, 갔 다 않, 갔 다 애, 갔 다 야, 갔 다 어, 갔 다 어디, 갔 다 어제, 갔 다 언니, 갔 다 얼마, 갔 다 엄마, 갔 다 엄청, 갔 다 에, 갔 다 에서, 갔 다 여기, 갔 다 여긴, 갔 다 여행, 갔 다 역시, 갔 다 영화, 갔 다 옆, 갔 다 예약, 갔 다 예전, 갔 다 옛날, 갔 다 오늘, 갔 다 오랜만, 갔 다 오빠, 갔 다 온, 갔 다 온다, 갔 다 온다고, 갔 다 온지, 갔 다 올, 갔 다 올까, 갔 다 올께, 갔 다 올해, 갔 다 옴, 갔 다 와, 갔 다 와도, 갔 다 와서, 갔 다 와야, 갔 다 완전, 갔 다 왓, 갔 다 왔, 갔 다 왔었, 갔 다 왔으니, 갔 다 왔음, 갔 다 왜, 갔 다 왜냐면, 갔 다 요즘, 갔 다 우린, 갔 다 원래, 갔 다 위치, 갔 다 유명, 갔 다 을, 갔 다 이거, 갔 다 이건, 갔 다 이곳, 갔 다 이날, 갔 다 이때, 갔 다 이렇게, 갔 다 이름, 갔 다 이미, 갔 다 이번, 갔 다 이유, 갔 다 이제, 갔 다 인, 갔 다 인천, 갔 다 일단, 갔 다

강조 기 위해, 강조 다, 강조 된, 강조 듯, 강조 를, 강조 며, 강조 면서, 강조 셨, 강조 시, 강조 신, 강조 어, 강조 여, 강조 였, 강조 이유, 강조 지만, 강조 한다, 강조 할, 강조 합니다, 강조 해, 강조 해도, 강조 해도 지나치, 강조 해서, 강조 했, 강조 했 다, 강조 했 던, 강좌 도, 강좌 를, 강주 막, 강줄기 를, 강즙 을, 강지 도, 강진 에, 강진 에서, 강진 청자, 강진군 장흥군, 강진군 장흥군 여수시, 강진면 관촌면, 강진면 관촌면 덕치면, 강진읍 군동면, 강진읍 군동면 대구면, 강차 마석, 강차 마석 힐즈, 강창구 찹쌀, 강창구 찹쌀 진, 강창구 찹쌀 진순, 강천면 금사면, 강천면 금사면 능서면, 강철 비, 강철 체력, 강청 골, 강청 골 순대, 강촌 닭갈비, 강촌 레일바이크, 강촌 스키장, 강촌 식당, 강촌 에, 강촌 에서, 강추 ㅋ, 강추 ㅋㅋ, 강추 ㅋㅋ ㅋㅋ, 강추 ㅋㅋㅋ, 강추 ㅋㅋㅋ ㅋㅋ, 강추 ㅋㅋㅋ ㅋㅋㅋ, 강추 ㅎㅎ, 강추 ㅎㅎㅎ, 강추 가격, 강추 강추, 강추 강추 강추, 강추 강추 입니다, 강추 강추 합니다, 강추 강추 해요, 강추 건, 강추 고, 강추 고 곳, 강추 고 네요, 강추 고 다, 강추 고 습니다, 강추 고 어요, 강추 고 픈, 강추 고기, 강추 곳, 강추 곳 에요, 강추 곳 입니다, 강추 구요, 강추 길래, 강추 까지, 강추 꼭, 강추 너무, 강추 네요, 강추 니, 강추 다, 강추 다음, 강추 던, 강추 도, 강추 돈, 강추 드려, 강추 드려 용, 강추 드려요, 강추 드리, 강추 드리 고, 강추 드립, 강추 드립 니, 강추 드립니다, 강추 드립니다 도, 강추 드립니다 소, 강추 라고, 강추 랍니다, 강추 로, 강추 를, 강추 마지막, 강추 만, 강추 맛, 강추 맛있, 강추 맛집, 강추 먹, 강추 메뉴, 강추 메뉴 에, 강추 메뉴 입니다, 강추 며, 강추 바, 강추 밥, 강추 셔서, 강추 셨, 강추 수원, 강추 시, 강추 식당, 강추 신, 강추 아니, 강추 아이, 강추 아주, 강추 아주 맛

In [9]:
get_review_result2 = review_anal.get_review('./post_crawling_통합_정제_2.csv')

In [14]:
get_review_result1 = review_anal.get_review('./post_crawling_통합_정제_1.csv')

In [15]:
len(get_review_result1), len(get_review_result2)

(400000, 390465)

In [16]:
# review의 개수는 790,465 개 
get_review_result = get_review_result1 + get_review_result2
len(get_review_result)

790465

In [17]:
len("아 이정도면 한 문장이라고 할 수 있는 길이입니다. 그렇죠? 만약 아니라면.. 몰라요.")

48

In [19]:
# 한 문장의 길이는 대략 50이라고 가정하고 문장의 개수를 세겠습니다.

total_len = 0
for sentence in get_review_result:
    total_len += len(sentence)
total_len

1216401905

In [20]:
total_len/50

24328038.1

#### restaurant_review 데이터셋 설명 
- 790,465개 리뷰
- 24,328,038개 문장 
- (식당 개수, 식당 종류 등등 기타 정보 추가 예정) 

In [21]:
df1_names = pd.read_csv('./post_crawling_통합_정제_1_names.csv')
df2_names = pd.read_csv('./post_crawling_통합_정제_2_names.csv')

In [22]:
df1_names.head()

Unnamed: 0  names
0           0    ㄱ ㄱ
1           1  ㄱ ㄱ ㄱ
2           2  ㄱ ㄱ ㅅ
3           3  ㄱ ㄱ ㅆ
4           4  ㄱ ㄱ ㅋ

In [23]:
df2_names.head()

Unnamed: 0   names
0           0     ㄱ ㄱ
1           1   ㄱ ㄱ ㄱ
2           2   ㄱ ㄱ ㅅ
3           3   ㄱ ㄱ ㅋ
4           4  ㄱ ㄱ ㅋㅋ

In [24]:
df1_names.drop('Unnamed: 0', axis=1, inplace=True)
df2_names.drop('Unnamed: 0', axis=1, inplace=True)
df_names = pd.concat([df1_names, df2_names])
df_names.shape

(2665666, 1)

In [25]:
df_names.drop_duplicates(keep='first', inplace=True)
df_names.shape

(1528291, 1)

In [26]:
df_names.reset_index(drop=True,inplace=True)
df_names.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1528291 entries, 0 to 1528290
Data columns (total 1 columns):
names    1528291 non-null object
dtypes: object(1)
memory usage: 11.7+ MB


In [27]:
df_names.to_csv('reviews_morphs_mecab.csv', encoding='utf-8')